scripts/cast-on-tenderly/index.js

```diff
@@ -19,33 +20,91 @@ if (!SPELL_ADDRESS) {
-    const rpcUrl = `https://rpc.tenderly.co/fork/${forkId}`;
-    const forkUrl = `https://dashboard.tenderly.co/${process.env.TENDERLY_USER}/${process.env.TENDERLY_PROJECT}/fork/${forkId}`;
-    return { forkId, forkUrl, rpcUrl };
+const createTenderlyTestnet = async function (spellName) {
+    const slug = `${spellName.replace(/\s+/g, '-').toLowerCase()}-${randomUUID()}`;
+    const response = await makeTenderlyApiRequest({
+        method: 'post',
+        path: '/testnet/container',
+        body: {
+            slug,
+            displayName: spellName,
+            description: spellName,
+            networkConfig: {
+                networkId: NETWORK_ID,
+                blockNumber: 'latest',
+                baseFeePerGas: '1',
+                chainConfig: {
+                    chainId: NETWORK_ID,
+                },
+            },
+            explorerPage: 'DISABLED',
+            syncState: true,
+        },
+    });
+    const testnetId = response.data.container.id;
+    const rpcEndpointPublic = response.data.container.connectivityConfig.endpoints.find(
+        endpoint => endpoint.displayName === 'testnet'
+    );
+    const rpcEndpointPrivate = response.data.container.connectivityConfig.endpoints.find(
+        endpoint => endpoint.displayName === 'unlocked'
+    );
+    const explorerUrlPublic = `https://dashboard.tenderly.co/explorer/vnet/${rpcEndpointPublic.id}`;
+    return {
+        testnetId,
+        explorerUrlPublic,
+        rpcUrlPublic: rpcEndpointPublic.uri,
+        rpcUrlPrivate: rpcEndpointPrivate.uri,
+    };
 };
 
-const publishTenderlyTransaction = async function (forkId, transactionId) {
-    await makeTenderlyApiRequest(`/fork/${forkId}/transaction/${transactionId}/share`);
-    return `https://dashboard.tenderly.co/shared/fork/simulation/${transactionId}`;
+const publishTenderlyTestnet = async function (testnetId) {
+    const response = await makeTenderlyApiRequest({
+        method: 'put',
+        path: `/testnet/container/${testnetId}`,
+        body: {
+            explorerPage: 'ENABLED',
+        },
+    });
+    if (response.data?.container?.explorer_page !== 'ENABLED') {
+        throw new Error('failed to publish testnet');
+    }
 };
 
-const runSpell = async function () {
-    const { forkId, forkUrl, rpcUrl } = await createTenderlyFork();
-    console.info('private tenderly fork is created', forkUrl);
+const runSpell = async function (spellAddress) {
+    const spellName = await getSpellName(spellAddress);
+    console.info(`preparing to execute spell ${spellAddress} with name "${spellName}"`);
 
-    const provider = new ethers.providers.JsonRpcProvider(rpcUrl);
+    const { testnetId, explorerUrlPublic, rpcUrlPrivate } = await createTenderlyTestnet(spellName);
+    console.info(`tenderly testnet "${testnetId}" is created`);
+
+    const provider = new ethers.providers.JsonRpcProvider(rpcUrlPrivate);
     const signer = provider.getSigner();
```

Comment:
Suggestion: extract the part related to giving the hat to the spell and the casting into separate functions. Then all low-level details would be properly encapsulated, and the main function would be a sequence of well-defined steps:

1. Create a Tenderly Testnet
2. Give the hat to the spell
3. Cast the spell

Reply:
I would also normally put tenderly-related code, ethers-related and setup-related into a separate files, but thought it's against the convention of having a single huge file per script 😅

Reply:
Following up the latest review:

1. Create a Tenderly Testnet
2. ~~Give the hat to the spell~~ Cast the spell
3. ~~Cast the spell~~ Publish the Testnet